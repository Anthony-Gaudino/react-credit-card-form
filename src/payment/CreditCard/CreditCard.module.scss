@use "../../theme";

.creditCard {
  max-width: 450px;
  margin-right: auto;
  margin-left: auto;
  font-family: "Source Code Pro", monospace;
  color: #fff;
  text-shadow: 7px 6px 10px rgb(14 42 90 / 80%);
  cursor: pointer;
  perspective: 2000px;

  &__paper {
    position: relative;
    height: 270px;
    transition: transform 3s;
    transform-style: preserve-3d;

    &--flipped {
      transform: rotateY(180deg);
    }
  }

  &__surface {
    position: absolute;
    width: 100%;
    height: 100%;
    padding: 1.5rem;
    background: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' height='100%25' width='100%25'%3E%3Cdefs%3E%3Cpattern id='doodad' width='187' height='187' viewBox='0 0 40 40' patternUnits='userSpaceOnUse' patternTransform='rotate(24)'%3E%3Crect width='100%25' height='100%25' fill='rgba(24, 67, 140,1)'/%3E%3Cpath d='M5 20L20 0L20 40zM25 20L40 0L40 40z' fill='rgba(82, 205, 236,1)'/%3E%3Cpath d='M13 20L20 0L20 40zM33 20L40 0L40 40zM0-20L5 0L0 20zM0 20L5 40L0 60zM20-20L25 0L20 20zM20 20L25 40L20 60z' fill='rgba(228, 158, 62,1)'/%3E%3C/pattern%3E%3C/defs%3E%3Crect fill='url(%23doodad)' height='200%25' width='200%25'/%3E%3C/svg%3E ");
    border-radius: 8px;
    backface-visibility: hidden;

    &--back {
      transform: rotateY(180deg);
    }
  }

  &__contentWrapper {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    height: 100%;
  }

  &__row {
    display: flex;
    justify-content: space-between;

    &--right {
      justify-content: end;
    }
  }

  &__chip {
    width: 3rem;
    height: 2.5rem;
  }

  &__issuer {
    max-width: 100px;
    height: 60px;
    transform-style: preserve-3d;
  }

  &__number {
    font-size: 27px;
    font-weight: 500;
    line-height: 1;
    color: #fff;

    // Embossed text
    // text-shadow: 1px 1px rgba(230, 232, 236, 0.4), -2px -1px rgba(225, 225, 225, 0.4);
  }

  &__info {
    display: flex;
    flex-direction: column;
  }

  &__infoTitle {
    font-size: 13px;
    font-weight: 500;
    opacity: 0.7;
  }

  &__magneticTape {
    height: 4rem;
    margin: 0 -1.5rem;
    background: black;
  }

  &__CVVTitle {
    margin-right: 0.5rem;
    text-align: right;
  }

  &__CVVField {
    display: flex;
    justify-content: end;
    height: 3rem;
    padding: 0.5rem;
    color: #1a3b5d;
    background: white;
    border-radius: 8px;
  }
}
