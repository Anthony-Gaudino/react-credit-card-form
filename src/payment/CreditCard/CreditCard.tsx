import clsx from "clsx";

import Paper from "../../components/Paper";
import smartChip from "../../img/smart_chip.svg";
import styles from "./CreditCard.module.scss";

const getIssuer = (cardNumber: string): string => {
  const firstDigit = +cardNumber.substring(0, 1);
  const first2Digits = +cardNumber.substring(0, 2);
  const first3Digits = +cardNumber.substring(0, 3);
  const first4Digits = +cardNumber.substring(0, 4);
  const first6Digits = +cardNumber.substring(0, 6);

  // 6 Digits
  //=========================================================================
  if (
    (first6Digits >= 506099 && first6Digits <= 506198) ||
    (first6Digits >= 650002 && first6Digits <= 650027) ||
    (first6Digits >= 507865 && first6Digits <= 507964)
  ) {
    return "verve";
  }

  if ([676770, 676774].includes(first6Digits)) {
    return "maestro";
  }

  if (first6Digits === 417500) {
    return "visaelectron";
  }

  if (first6Digits >= 622126 && first6Digits <= 622925) {
    return "unionpay";
  }

  // 4 Digits
  //=========================================================================

  if (
    [5018, 5020, 5038, 5893, 6304, 6759, 6761, 6762, 6763].includes(
      first4Digits
    )
  ) {
    return "maestro";
  }

  if ([4026, 4508, 4844, 4913, 4917].includes(first4Digits)) {
    return "visaelectron";
  }

  if (first4Digits >= 3528 && first4Digits <= 3589) {
    return "jcb";
  }

  if (first4Digits === 6011) {
    return "discover";
  }

  if (first4Digits >= 2221 && first4Digits <= 2720) {
    return "mastercard";
  }

  if (first4Digits === 9792) {
    return "troy";
  }

  // 3 Digits
  //=========================================================================

  if (first3Digits >= 644 && first3Digits <= 649) {
    return "discover";
  }

  // 2 Digits
  //=========================================================================

  if ([34, 37].includes(first2Digits)) {
    return "amex";
  }

  if ([31, 62].includes(first2Digits)) {
    return "unionpay";
  }

  if ([36, 54].includes(first2Digits)) {
    return "diners";
  }

  if (first2Digits === 65) {
    return "discover";
  }

  if (first2Digits >= 51 && first2Digits <= 55) {
    return "mastercard";
  }

  if (first2Digits === 65) {
    return "troy";
  }

  // 1 Digit
  //=========================================================================
  if (firstDigit === 4) {
    return "visa";
  }

  // Unknown
  //=========================================================================
  return "other";
};

interface CreditCardProps {
  /**
   * Card owner name as it's printed on card.
   */
  name: string;
  /**
   * The card number without any spaces between numbers.
   */
  cardNumber: string;
  /**
   * The card number as it's printed on card.
   */
  printedCardNumber: string;
  /**
   * The card's expiration month.
   */
  month: string;
  /**
   * The card's expiration year.
   */
  year: string;
  /**
   * The card's CVV code.
   */
  cvv: string;
  /**
   * If `true` shows the back of the card.
   */
  showBack?: boolean;
  /**
   * Callback for card click or keypress.
   */
  cardClickHandler: () => void;
}

/**
 * A credit card widged that can be flipped.
 */
export const CreditCard = (props: CreditCardProps): JSX.Element => {
  const {
    name,
    cardNumber,
    printedCardNumber,
    month,
    year,
    cvv,
    showBack = false,
    cardClickHandler,
  } = props;

  const issuer = getIssuer(cardNumber);
  const issuerLogo = require(`../../img/logos/${issuer}.svg`);
  const cardMonth = month ? month : "MM";
  const cardYear = year ? year.slice(2) : "YY";

  const classes = clsx({
    [styles.creditCard__paper]: true,
    [styles["creditCard__paper--flipped"]]: showBack,
  });

  const onCardClickHandler = () => cardClickHandler();

  const cardKeyPressHandler = (e: React.KeyboardEvent<HTMLDivElement>) => {
    if (e.key === "Enter") cardClickHandler();
  };

  return (
    <div
      className={styles.creditCard}
      onClick={onCardClickHandler}
      role="button"
      tabIndex={-1}
      onKeyPress={cardKeyPressHandler}
    >
      <Paper elevation={6} className={classes}>
        <div
          className={`${styles.creditCard__surface} ${styles["creditCard__surface--front"]}`}
        >
          <div className={styles.creditCard__contentWrapper}>
            <div className={styles.creditCard__row}>
              <img className={styles.creditCard__chip} src={smartChip} alt="" />
              <img
                className={styles.creditCard__issuer}
                src={issuerLogo}
                alt=""
              />
            </div>
            <div className={styles.creditCard__row}>
              <span className={styles.creditCard__number}>
                {printedCardNumber}
              </span>
            </div>
            <div className={styles.creditCard__row}>
              <div className={styles.creditCard__info}>
                <div className={styles.creditCard__infoTitle}>Card Holder</div>
                <div className={styles.creditCard__infoContent}>{name}</div>
              </div>
              <div className={styles.creditCard__info}>
                <div className={styles.creditCard__infoTitle}>Expires</div>
                <div className={styles.creditCard__infoContent}>
                  {cardMonth}/{cardYear}
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className={`${styles.creditCard__surface} ${styles["creditCard__surface--back"]}`}
        >
          <div className={styles.creditCard__contentWrapper}>
            <div className={styles.creditCard__magneticTape} />
            <div className={styles.creditCard__signatureAndCVVWrapper}>
              <div className={styles.creditCard__CVVTitle}>CVV</div>
              <div className={styles.creditCard__CVVField}>
                <span>{cvv}</span>
              </div>
            </div>
            <div
              className={`${styles.creditCard__row} ${styles["creditCard__row--right"]}`}
            >
              <img
                className={styles.creditCard__issuer}
                src={issuerLogo}
                alt=""
              />
            </div>
          </div>
        </div>
      </Paper>
    </div>
  );
};
