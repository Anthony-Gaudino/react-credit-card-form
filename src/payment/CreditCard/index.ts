export { CreditCard as default } from "./CreditCard";
