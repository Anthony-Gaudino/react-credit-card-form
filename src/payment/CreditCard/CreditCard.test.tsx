import { render, screen } from "@testing-library/react";

import CreditCard from ".";

describe("Test CreditCard component", () => {
  test("the widget is correctly rendered", () => {
    const { asFragment } = render(
      <CreditCard
        name="John Doe"
        cardNumber="4000000000000000"
        printedCardNumber="4000 0000 0000 0000"
        month="01"
        year="2025"
        cvv="123"
        cardClickHandler={() => {}}
      />
    );

    const name = screen.getByText("John Doe");
    const cardNumber = screen.getByText("4000 0000 0000 0000");
    const expireDate = screen.getByText("01/25");
    const cvv = screen.getByText("123");

    expect(name).toBeInTheDocument();
    expect(cardNumber).toBeInTheDocument();
    expect(expireDate).toBeInTheDocument();
    expect(cvv).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });
});
