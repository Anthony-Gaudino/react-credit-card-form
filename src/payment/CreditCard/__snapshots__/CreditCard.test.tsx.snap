// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CreditCard component the widget is correctly rendered 1`] = `
<DocumentFragment>
  <div
    class="creditCard"
    role="button"
    tabindex="-1"
  >
    <div
      class="paper creditCard__paper"
      style="--elevation: 6;"
    >
      <div
        class="creditCard__surface creditCard__surface--front"
      >
        <div
          class="creditCard__contentWrapper"
        >
          <div
            class="creditCard__row"
          >
            <img
              alt=""
              class="creditCard__chip"
              src="smart_chip.svg"
            />
            <img
              alt=""
              class="creditCard__issuer"
              src="[object Object]"
            />
          </div>
          <div
            class="creditCard__row"
          >
            <span
              class="creditCard__number"
            >
              4000 0000 0000 0000
            </span>
          </div>
          <div
            class="creditCard__row"
          >
            <div
              class="creditCard__info"
            >
              <div
                class="creditCard__infoTitle"
              >
                Card Holder
              </div>
              <div
                class="creditCard__infoContent"
              >
                John Doe
              </div>
            </div>
            <div
              class="creditCard__info"
            >
              <div
                class="creditCard__infoTitle"
              >
                Expires
              </div>
              <div
                class="creditCard__infoContent"
              >
                01/25
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="creditCard__surface creditCard__surface--back"
      >
        <div
          class="creditCard__contentWrapper"
        >
          <div
            class="creditCard__magneticTape"
          />
          <div
            class="creditCard__signatureAndCVVWrapper"
          >
            <div
              class="creditCard__CVVTitle"
            >
              CVV
            </div>
            <div
              class="creditCard__CVVField"
            >
              <span>
                123
              </span>
            </div>
          </div>
          <div
            class="creditCard__row creditCard__row--right"
          >
            <img
              alt=""
              class="creditCard__issuer"
              src="[object Object]"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
